src/Controls/src/Core/VisualElement.cs

```diff
@@ -249,44 +250,78 @@ static void OnTransformChanged(BindableObject bindable, object oldValue, object
 					(bindable as VisualElement)?.NotifyBackgroundChanges();
 			});
 
+		readonly WeakBackgroundChangedProxy _backgroundProxy = new();
+
+		~VisualElement() => _backgroundProxy.Unsubscribe();
+
 		void NotifyBackgroundChanges()
 		{
-			if (Background is ImmutableBrush)
+			var background = Background;
+			if (background is ImmutableBrush)
 				return;
 
-			if (Background != null)
+			if (background != null)
 			{
-				Background.Parent = this;
-				Background.PropertyChanged += OnBackgroundChanged;
-
-				if (Background is GradientBrush gradientBrush)
-					gradientBrush.InvalidateGradientBrushRequested += InvalidateGradientBrushRequested;
+				SetInheritedBindingContext(background, BindingContext);
+				_backgroundProxy.Subscribe(background, (sender, e) => OnPropertyChanged(nameof(Background)));
 			}
 		}
 
 		void StopNotifyingBackgroundChanges()
 		{
-			if (Background is ImmutableBrush)
+			var background = Background;
+			if (background is ImmutableBrush)
 				return;
 
-			if (Background != null)
+			if (background != null)
 			{
-				Background.Parent = null;
-				Background.PropertyChanged -= OnBackgroundChanged;
-
-				if (Background is GradientBrush gradientBrush)
-					gradientBrush.InvalidateGradientBrushRequested -= InvalidateGradientBrushRequested;
+				SetInheritedBindingContext(background, null);
+				_backgroundProxy.Unsubscribe();
 			}
 		}
 
-		void OnBackgroundChanged(object sender, PropertyChangedEventArgs e)
+		class WeakBackgroundChangedProxy : WeakEventProxy<Brush, PropertyChangedEventHandler>
 		{
-			OnPropertyChanged(nameof(Background));
-		}
+			void OnBackgroundChanged(object sender, PropertyChangedEventArgs e)
+			{
+				if (TryGetHandler(out var handler))
+				{
+					handler(sender, e);
+				}
+				else
+				{
+					Unsubscribe();
+				}
+			}
 
-		void InvalidateGradientBrushRequested(object sender, EventArgs e)
-		{
-			OnPropertyChanged(nameof(Background));
+			public override void Subscribe(Brush source, PropertyChangedEventHandler handler)
+			{
+				if (TryGetSource(out var s))
+				{
+					s.PropertyChanged -= OnBackgroundChanged;
+
+					if (s is GradientBrush g)
+						g.PropertyChanged -= OnBackgroundChanged;
+				}
+
+				source.PropertyChanged += OnBackgroundChanged;
+				if (source is GradientBrush gradientBrush)
+					gradientBrush.PropertyChanged += OnBackgroundChanged;
```

Comment:
This appears to be a double subscription? Should this have been the other event? If so, maybe we don't need the generic property event if this is a gradient?

Reply:
`InvalidateGradientBrushRequested` looks to be fired when `CollectionChanged` fires from `GradientStopCollection`.